Add nRF52 SPI master (SPIM) driver.

JayKickliter opened the pull request:
### Pull Request Overview

This PR adds a SPI master driver to the nrf52 crate. It uses SPIM peripherals with EasyDMA. It does not use the non-DMA SPI peripheral.

### Testing Strategy

Testing requires modifying the `nRF52DK` board crate to use the driver. I have a separate branch that does this and has been tested with `userland/examples/tests/rf233`.

### TODO or Help Wanted

I have not implemented any of single-byte read/write operations. It looks like they aren't used anywhere in the code base. If I'm wrong, and they are required, I can add them.

### Documentation Updated

- [x] Kernel: The relevant files in `/docs` have been updated or **no updates are required**.
- [x] Userland: The application README has been added, updated, or **no updates are required**.

### Formatting

- [x] `make formatall` has been run.

### Working Example

Configuring the ST S2-LP radio:

```c
void
s2lp_write_registers(uint8_t addr, uint8_t const * data, size_t data_len)
{
    uint8_t write_buf[64] = {0, addr};
    uint8_t read_buf[64]  = {0, addr};
    memcpy(&write_buf[2], data, data_len);
    int rc = spi_read_write_sync((const char *)write_buf,
                                 (char *)read_buf,
                                 data_len + 2);
    CHECK_ERR(rc);
}

void
s2lp_init(void)
{
    uint8_t tmp[6];
    tmp[0] = 0x40; /* reg. FIFO_CONFIG3 (0x3C) */
    tmp[1] = 0x40; /* reg. FIFO_CONFIG2 (0x3D) */
    tmp[2] = 0x40; /* reg. FIFO_CONFIG1 (0x3E) */
    tmp[3] = 0x40; /* reg. FIFO_CONFIG0 (0x3F) */
    tmp[4] = 0x41; /* reg. PCKT_FLT_OPTIONS (0x40) */
    s2lp_write_registers(0x3C, tmp, 5);
}
```

![screenshot 2017-09-07 14 52 53](https://user-images.githubusercontent.com/2551201/30186989-aec4e048-93dc-11e7-8316-742266396d88.png)

ppannuto commented on chips/nrf52/src/spi.rs:
```diff
@@ -0,0 +1,501 @@
+use chip;
```
We usually like a module comment at the top (e.g. https://github.com/helena-project/tock/blob/master/chips/sam4l/src/adc.rs#L1). Things worth including:
  - what is it (usually an obvious one-liner)
  - how complete is the implementation / any caveats
  - an author line
  - a last (major) update date

JayKickliter replied:
Done.

ppannuto commented on chips/nrf52/src/spi.rs:
```diff
@@ -0,0 +1,501 @@
+pub static mut SPIM1: SPIM = SPIM::new(1);
+pub static mut SPIM2: SPIM = SPIM::new(2);
+
+mod registers {
```
We're also (trying to) more aggressively include `///` comments towards the generation of nice documentation (see https://rustdoc--tockosorg.netlify.com/rustdoc/sam4l/adc/struct.AdcChannel.html for the working branch)

I think some simple 'what is this' comments attached to all the structures would be great - also most of the `//`'s below could probably upgrade to `///` pretty happily. The SAM4L ADC implementation linked about is a decent template for the balance we're going for (though definitely open to feedback/pushback on documentation)

ppannuto commented on chips/nrf52/src/spi.rs:
```diff
@@ -0,0 +1,501 @@
+            // End of RXD buffer and TXD buffer reached
+            match self.chip_select.get() {
+                Some(cs) => cs.set(),
+                None => panic!("Chip select not configured."),
```
In general, we try to avoid `panic`ing if we can help it -- on the downcall you returned `ENODEVICE` which makes a lot of sense. It's not immediately obvious to me how you would even reach the callback w/out having set a chip select (in practice) here, so maybe it's okay.

Perhaps just adding a comment here that explains when this `panic` (should) never trigger in practice would alleviate my concerns. Perhaps it makes sense to convert this to a `debug_assert` and just bail out in the production case. Unsure..

JayKickliter commented:
@ppannuto thanks for the feedback. I'll address the individual requests todayish.

niklasad1 commented on chips/nrf52/src/spi.rs:
```diff
@@ -0,0 +1,501 @@
+            Enabled = 7,
+        }
+
+        #[allow(non_snake_case)]
```
Seems like this is unnecessary since all is already snake_cases?

JayKickliter replied:
Good catch. I started out by copy-pasting some other code as a template and that must have came with it.

niklasad1 commented:
Very nice work, however is not [boards/nrf52dk](https://github.com/helena-project/tock/blob/master/boards/nrf52dk/src/main.rs) supposed to be changed?  (i.e., allocate 4 specific GPIO pins or something like that)

JayKickliter commented:
@niklasad1 maybe, but what pins? It's kind of application specific. In my case, we'd be making a custom board any not using the nrf52dk board crate.

niklasad1 commented:
ohh very cool, if you want to it add for nRF52-DK you can pick 4 pins that are not currently in use i.e., some are reserved for LEDs, Buttons, UART etc but we have documented it the nRF52-DK crate.

JayKickliter commented:
Adding WIP tag. In an effort to make some code more readable right before opening this PR, I somehow managed to break it so that the read SPI data doesn't make it to userland.

JayKickliter commented:
It turns out the SPI driver is calling the client callback with the correct data. I don't know why it's not making its way up to userland.

JayKickliter commented:
It turns out I deleted a line that set the transfer length. All is good now. I removed WIP from the PR title, but I still need to address review comments.

JayKickliter commented:
Driver in action showing SPI data making its way to a userland c app:

[![asciicast](https://asciinema.org/a/1p6ci7PVIQ8mOebgPkQzIyhMs.png)](https://asciinema.org/a/1p6ci7PVIQ8mOebgPkQzIyhMs)

And the console output:
```
/********************************************/
/* App started                              */
0x00 0x00 0x48 
/* App finished                             */
/********************************************/
```

alevy commented:
@JayKickliter looks to me like you addressed @ppannuto's comments. And looks good to me. @ppannuto, we good to merge?

ppannuto commented:
Awesome to see this land -- thanks Jay!
